rs_utils.py

```diff
@@ -82,6 +84,324 @@ def close(self):
 
 
+class RSNamespace(argparse.Namespace):
+    def __getitem__(self, key):
+        try:
+            return self.__getattribute__(key)
+        except AttributeError:
+            raise KeyError
+
+    def __setitem__(self, key, value):
+        self.__setattr__(key, value)
+
+
+class RSArgumentParser(argparse.ArgumentParser):
+
+    def __init__(self, add_help=False, ignore_unrecognized_arguments=True, **kwargs):
+        self.ignore_unrecognized_arguments = ignore_unrecognized_arguments
+        super().__init__(self, add_help=add_help, **kwargs)
+
+    def exit(self, status=0, message=None):
+        raise CommandSyntaxError(message)
+
+    def error(self, message):
+        """error(message: string)
+
+        Prints a usage message incorporating the message to stderr and
+        exits.
+
+        If you override this in a subclass, it should not return -- it
+        should either exit or raise an exception.
+        """
+        # self.print_usage(_sys.stderr)
+        # args = {'prog': self.prog, 'message': message}
+        # self.exit(2, _('%(prog)s: error: %(message)s\n') % args)
+        raise CommandSyntaxError(message)
+
+    def parse_args(self, args=None, namespace=None):
+        args, argv = self.parse_known_args(args, namespace)
+        if argv and not self.ignore_unrecognized_arguments:
+            msg = 'Unrecognized arguments: %s'
+            self.error(msg % ' '.join(argv))
+        return args
+
+    def parse_known_args(self, args=None, namespace=None):
+        if args is None:
+            raise CommandSyntaxError("No arguments. How did you even manage that.")
+        else:
+            # make sure that args are mutable
+            args = list(args)
+
+        # default Namespace built from parser defaults
+        if namespace is None:
+            namespace = RSNamespace()
+
+        # add any action defaults that aren't present
+        for action in self._actions:
+            if action.dest is not argparse.SUPPRESS:
+                if not hasattr(namespace, action.dest):
+                    if action.default is not argparse.SUPPRESS:
+                        setattr(namespace, action.dest, action.default)
+
+        # add any parser defaults that aren't present
+        for dest in self._defaults:
+            if not hasattr(namespace, dest):
+                setattr(namespace, dest, self._defaults[dest])
+
+        # parse the arguments and exit if there are any errors
+        try:
+            namespace, args = self._parse_known_args(args, namespace)
+            if hasattr(namespace, argparse._UNRECOGNIZED_ARGS_ATTR):
+                args.extend(getattr(namespace, argparse._UNRECOGNIZED_ARGS_ATTR))
+                delattr(namespace, argparse._UNRECOGNIZED_ARGS_ATTR)
+            return namespace, args
+        except argparse.ArgumentError:
+            err = argparse._sys.exc_info()[1]
+            self.error(str(err))
+
+    def _parse_known_args(self, arg_strings, namespace):
```

Comment:
Is copying all of this actually necessary?

Reply:
```py
        # default Namespace built from parser defaults
        if namespace is None:
            namespace = RSNamespace()
```
Yes, because of this bit. Tried just putting that in with `super()`, didn't work well.

Reply:
I suppose I'll play with it a bit later.